### PR TITLE
[perf] Use direct copy (broadcast) instead of cat for k_nope/k_pe in MLA prefill

### PR DESCRIPTION
**Pull request by minosfuture**
## Purpose

Reduce k tensor concatenation latency, from **3.16ms to 1.61ms** for batch size 32768 (i.e., `k.shape=torch.Size([32768, 128, 192]), k_nope.shape=torch.Size([32768, 128, 128]), k_pe.shape=torch.Size([32768, 1, 64])`.

image

## Test Plan

* verified perf improvement with trace
* verified accuracy
* microbenchmark: `python benchmarks/kernels/benchmark_mla_k_concat.py`

## Test Result

local-completions (model=nvidia/DeepSeek-R1-0528-FP4,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=32), gen_kwargs: (None), limit: 100.0, num_fewshot: None, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.96|±  |0.0197|
|     |       |strict-match    |     5|exact_match|↑  | 0.96|±  |0.0197|

```
================================================================================
Benchmark: torch.cat vs direct copy for MLA k_nope/k_pe concatenation
================================================================================
Tensor shapes: k_nope=[B, 128, 128], k_pe=[B, 1, 64]
dtype: float8_e4m3fn

  Batch Size |   cat (ms) |  direct (ms) |  Speedup |  Reduction
----------------------------------------------------------------------
          32 |      0.010 |        0.027 |    0.38x |    -160.6%
          64 |      0.010 |        0.028 |    0.37x |    -169.9%
         128 |      0.015 |        0.027 |    0.53x |     -87.0%
         256 |      0.027 |        0.027 |    0.98x |      -1.9%
         512 |      0.049 |        0.027 |    1.82x |      45.0%
        1024 |      0.094 |        0.044 |    2.13x |      53.1%
        2048 |      0.185 |        0.086 |    2.14x |      53.2%
        4096 |      0.401 |        0.205 |    1.95x |      48.8%
        8192 |      0.797 |        0.405 |    1.97x |      49.1%
       16384 |      1.583 |        0.801 |    1.97x |      49.4%
       32768 |      3.159 |        1.593 |    1.98x |      49.6%
       65536 |      6.308 |        3.176 |    1.99x |      49.7%
================================================================================

Speedup summary:
  Min:  0.37x
  Max:  2.14x
  Mean: 1.52x

Conclusion:
  - Direct copy becomes beneficial at batch size >= 512
  - For batch sizes >= 512: avg speedup = 1.99x
  - MLA prefill typically uses large batches, so optimization is effective
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Review by tlrmchlsmth — COMMENTED** [review]
Have you profiled with shorter batch sizes as well? Is this faster across the board? Any cases where the performance is slower?

**Comment by minosfuture** [comment]
> Reduce k tensor concatenation latency, from **3.16ms to 1.61ms** for batch size 32768 (i.e., `k.shape=torch.Size([32768, 128, 192]), k_nope.shape=torch.Size([32768, 128, 128]), k_pe.shape=torch.Size([32768, 1, 64])`.

nice table from a bench script by claude:

     ================================================================================
     Benchmark: torch.cat vs direct copy for MLA k_nope/k_pe concatenation
     ================================================================================
     Tensor shapes: k_nope=[B, 128, 128], k_pe=[B, 1, 64]
     dtype: float8_e4m3fn

       Batch Size |   cat (ms) |  direct (ms) |  Speedup |  Reduction
     ----------------------------------------------------------------------
               32 |      0.010 |        0.026 |    0.39x |    -154.9%
               64 |      0.010 |        0.027 |    0.38x |    -162.0%
              128 |      0.015 |        0.026 |    0.55x |     -81.4%
              256 |      0.027 |        0.026 |    1.01x |       1.4%
              512 |      0.049 |        0.026 |    1.86x |      46.3%
             1024 |      0.094 |        0.044 |    2.15x |      53.4%
             2048 |      0.185 |        0.086 |    2.14x |      53.3%
             4096 |      0.401 |        0.206 |    1.95x |      48.7%
             8192 |      0.797 |        0.405 |    1.96x |      49.1%
            16384 |      1.583 |        0.801 |    1.98x |      49.4%
            32768 |      3.159 |        1.593 |    1.98x |      49.6%
            65536 |      6.309 |        3.176 |    1.99x |      49.7%
     ================================================================================

     Speedup summary:
       Min:  0.38x
       Max:  2.15x
       Mean: 1.53x

     Conclusion:
       - Direct copy becomes beneficial at batch size >= 256
       - For batch sizes >= 512: avg speedup = 2.00x
       - MLA prefill typically uses large batches, so optimization is effective

10us difference with small batch size seems negligible. I'd trade it for simpler code.  Wdyt? @tylertitsworth

**Review by tlrmchlsmth — APPROVED** [review]
Thank you for running that benchmark!

File: vllm/v1/attention/backends/mla/common.py [hunk 1]
```diff
@@ -1645,6 +1645,33 @@ def get_and_maybe_dequant_weights(layer: LinearBase):
...
+            Tensor of shape [..., nope_dim + pe_dim]
+        """
+        k = torch.empty(
+            (*k_nope.shape[:-1], k_nope.shape[-1] + k_pe.shape[-1]),
+            dtype=k_nope.dtype,
+            device=k_nope.device,
+        )
+        # Direct copies with efficient broadcasting
+        k[..., : k_nope.shape[-1]] = k_nope
+        k[..., k_nope.shape[-1] :] = k_pe
```
Comment (mgoin) [inline vllm/v1/attention/backends/mla/common.py:1681]: Could using `.copy_(` be any faster?

**Review by mgoin — APPROVED** [review]
LGTM, just one possible improvement

File: vllm/v1/attention/backends/mla/common.py (hunk 1 above)
Comment (minosfuture) [reply vllm/v1/attention/backends/mla/common.py:1681]: lemme check

Comment (minosfuture) [reply vllm/v1/attention/backends/mla/common.py:1681]: it is copy underlying:
image
